db/schema.rb

```diff
@@ -372,8 +372,9 @@
     t.string   "state",           limit: 255
     t.integer  "order_id"
     t.boolean  "included",                                             default: false
-    t.index ["adjustable_id"], name: "index_adjustments_on_order_id", using: :btree
+    t.index ["adjustable_type", "adjustable_id"], name: "index_spree_adjustments_on_adjustable_type_and_adjustable_id", using: :btree
     t.index ["order_id"], name: "index_spree_adjustments_on_order_id", using: :btree
+    t.index ["originator_type", "originator_id"], name: "index_spree_adjustments_on_originator_type_and_originator_id", using: :btree
```

Comment:
I'm wondering if there is any scenario in which we don't filter by both type and id. If there was any, let's say we used `originator_id` alone, PostgreSQL won't be able to use any of these two indexes. See https://www.postgresql.org/docs/9.1/indexes-multicolumn.html

> A multicolumn B-tree index can be used with query conditions that involve any subset of the index's columns, but the index is most efficient when there are constraints on the leading (leftmost) columns. 

tl;dr; having a `WHERE originator_type ...` this index will be picked up, whereas with `WHERE orginator_id` it won't.

Nonetheless, I think the gains of having the `_type` in the compound index might be marginal. What's important here is to index by those two ids because the type ones have very [low cardinality](https://en.wikipedia.org/wiki/Cardinality_(SQL_statements)). Have you tried even without them?

Reply:
>PostgreSQL won't be able to use any of these two indexes

I read it as: Postgres can use both indexes in that case, but it would be slightly less efficient?

We very commonly use `orginator_type` for filtering (without the id), all the main adjustments scopes are using things like `originator_type: 'EnterpriseFee'`.

Reply:
I'll take another look :+1:

Reply:
Alright, looking at three options (before these changes, adding just the `*_id` indexes, and adding the compound indexes with `[*_type, *_id]`) there are some common queries that come out the same in each case, here's what `order.all_adjustments.enterprise_fee` looks like:
```
                                                   QUERY PLAN
----------------------------------------------------------------------------------------------------------------
 Index Scan using index_spree_adjustments_on_order_id on spree_adjustments  (cost=0.42..21.06 rows=4 width=140)
   Index Cond: (order_id = 376557)
   Filter: ((originator_type)::text = 'EnterpriseFee'::text)
(3 rows)
```
It's the same for all three. 

In other cases there's a big difference. Let's try `order.line_item_adjustments.tax` as an example. If we don't use the compound indexes with `type`, we get this:
```
                                                             QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=237.36..237.37 rows=1 width=140)
   Sort Key: spree_adjustments.created_at
   ->  Nested Loop  (cost=0.84..237.35 rows=1 width=140)
         ->  Index Scan using index_line_items_on_order_id on spree_line_items  (cost=0.42..13.02 rows=8 width=4)
               Index Cond: (order_id = 376557)
         ->  Index Scan using index_spree_adjustments_on_adjustable_id on spree_adjustments  (cost=0.42..28.03 rows=1 width=140)
               Index Cond: (adjustable_id = spree_line_items.id)
               Filter: (((adjustable_type)::text = 'Spree::LineItem'::text) AND ((originator_type)::text = 'Spree::TaxRate'::text))
(8 rows)
```

But when we add the compound indexes we get this:
```
                                                                     QUERY PLAN
----------------------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=18.67..18.67 rows=1 width=140)
   Sort Key: spree_adjustments.created_at
   ->  Nested Loop  (cost=0.84..18.66 rows=1 width=140)
         ->  Index Scan using index_spree_adjustments_on_adjustable_type_and_adjustable_id on spree_adjustments  (cost=0.42..8.44 rows=1 width=140)
               Index Cond: ((adjustable_type)::text = 'Spree::LineItem'::text)
               Filter: ((originator_type)::text = 'Spree::TaxRate'::text)
         ->  Index Scan using spree_line_items_pkey on spree_line_items  (cost=0.42..8.44 rows=1 width=4)
               Index Cond: (id = spree_adjustments.adjustable_id)
               Filter: (order_id = 376557)
(9 rows)
```